docs/man/fnb_judge.8.in
.\" fnb_judge - Fast Naive Bayes command line classifier (judge)
.TH "fnb_judge" "8" "Aug 2013"  "Trever Adams" ""
.SH "NAME"
fnb_judge \- Fast Naive Bayes command line classifier (judge)
.SH "SYNOPSIS"
\fBfnb_judge\fP -p \fIPRIMARY_HASH_SEED\fP -s \fISECONDARY_HASH_SEED\fP -i \fIINPUT_FILE_TO_JUDGE\fP -d \fICATEGORY_FNB_FILES_DIR\fP
.SH "DESCRIPTION"
.PP
\fBfnb_judge\fP is a command-line, stand alone classifier used to test fnb
classifiers. fnb stands for Fast Naive Bayes.
.PP

.PP
.SH "OPTIONS"
.PP
.BR PRIMARY_HASH_SEED
.br
.BR SECONDARY_HASH_SEED
.PP
   The hash seeds are used to prime the hashing algorithm. If
   these do not match in both the trainer, when training the
   data, and the classifier when classifying, the results are
   undefined. For this reason, this can be used as a sort of
   pseudo copy protection.
.PP
   Additionally, these values determine, to an extent, the
   number of collisions in hash data. The only way to know if
   you have chosen a good value is testing.
.PP
   Each seed is a hexadecimal representation of a 32 bit number.
   If is \fBnot\fR prefixed with "0x". All alpha characters
   should be uppercase.
.PP
   These \fBmust\fR be the same for all categories.
.PP
.BR INPUT_FILE_TO_JUDGE
.PP
   This is the file to judge, of course. It must be a UTF-8
   text file, HTML file, or XML file.
.PP
.BR CATEGORY_FNB_FILES_DIR
.PP
   This is the directory where the FNB files, aka FNB
   classifier data, are stored.
.PP
.BR RELATED_STRING
.PP
   This is the directory in which the candidate files are located.
   This is specially formated string made up of substrings.
   The substrigns are joined by "=".
   Each substring has the format of "\fIprimary\fR,\fIsecondary\fR,\fIbidirectional\fR".
.PP
   \fBPrimary\fR is the primary category name.
.PP
   \fBSecondary\fR is the secondary category name.
.PP
   \fBbidirectional\fR determines whether or not the similarity goes both
   directions. Not all should.
.PP
   An example of such a string is:
   "adult.affairs,social.dating,1=adult.*,adult.*,0=clothing.*,clothing.*,0=information.culinary,commerce.food,1"
.PP
   Please, note, this string \fBmust\fR be enclosed in double quotes.
   Also, this option (-r) cannot be provided more than once.
.PP
WARNING: Spaces and case matter.
.PP
.SH "NOTES"
You must always have all the files within each category share the hash seeds.
Additionally, all categories within the same classifier data set \fBmust\fR
share the same hash seeds.

.SH "FILES"
.nf
NONE
.fi

.PP
.SH "SEE ALSO"
.nf
.I fnb_findtolearn (8)
.I fnb_learn (8)
.I fnb_makepreload (8)
.fi

.PP
.SH "AUTHORS"
.nf
Trever Adams
.fi

.PP
.SH "BUGS"
There of course aren't any bugs, but if you find any, you should first
consult https://github.com/treveradams/C-ICAP-Classify and, if
necessary, file a bug report there.
.fi


services/classify/fnb_judge.c
/*
 *  Copyright (C) 2008-2012 Trever L. Adams
 *
 *  This file is part of srv_classify c-icap module and accompanying tools.
 *
 *  srv_classify is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as
 *  published by the Free Software Foundation, either version 3 of
 *  the License, or (at your option) any later version.
 *
 *  srv_classify is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
 */


#define _GNU_SOURCE

#ifndef _FILE_OFFSET_BITS
#define _FILE_OFFSET_BITS 64
#endif
#if (_FILE_OFFSET_BITS != 64)
#undef _FILE_OFFSET_BITS
#define _FILE_OFFSET_BITS 64
#endif

#define NOT_CICAP

#include <stdio.h>
#include <stdint.h>
#include <inttypes.h>
#include <limits.h>
#include <stdlib.h>
#include <string.h>
#include <locale.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <wchar.h>
#include <wctype.h>
#include <time.h>
#include <dirent.h>

#include "hash.c"
#include "bayes.c"
#include "train_common.h"

char *judge_file;
char *fbc_dir;

int readArguments(int argc, char *argv[])
{
int i;
char temp[PATH_MAX];

	if(argc < 9)
	{
		printf("Format of arguments is:\n");
		printf("\t-p PRIMARY_HASH_SEED\n");
		printf("\t-s SECONDARY_HASH_SEED\n");
		printf("\t-i INPUT_FILE_TO_JUDGE\n");
		printf("\t-d CATEGORY_FNB_FILES_DIR\n");
		printf("\t-r Related categories in form of \"primary,secondary,bidirectional\". Bidirectional should be 1 for yes, 0 for no. This option should only be supplied once. To include more than one, separate with \"=\".\n");
		printf("Spaces and case matter.\n");
		return -1;
	}
	for(i=1; i<9; i+=2)
	{
		if(strcmp(argv[i], "-p") == 0) sscanf(argv[i+1], "%"PRIx32, &HASHSEED1);
		else if(strcmp(argv[i], "-s") == 0) sscanf(argv[i+1], "%"PRIx32, &HASHSEED2);
		else if(strcmp(argv[i], "-i") == 0)
		{
			judge_file = malloc(strlen(argv[i+1]) + 1);
			sscanf(argv[i+1], "%s", judge_file);
		}
		else if(strcmp(argv[i], "-d") == 0)
		{
			int len = strlen(argv[i+1]);
			fbc_dir = malloc(len + 1);
			sscanf(argv[i+1], "%s", fbc_dir);
			if(fbc_dir[len-1] == '/') fbc_dir[len-1] = '\0';
		}
		else if(strcmp(argv[i], "-r") == 0)
		{
			strncpy(temp, argv[i+1], PATH_MAX-1);
			setupPrimarySecondFromCmdLine(temp);
		}
	}
/*	printf("Primary Seed: %"PRIX32"\n", HASHSEED1);
	printf("Secondary Seed: %"PRIX32"\n", HASHSEED2);
	printf("Learn File: %s\n", judge_file);*/
	return 0;
}

int main (int argc, char *argv[])
{
regexHead myRegexHead = {.head = NULL, .tail = NULL, .dirty = 0, . main_memory = NULL, .arrays = NULL, .lastarray = NULL};
wchar_t *myData;
HashList myHashes;
clock_t start, end;
HTMLClassification classification;
	checkMakeUTF8();
	initHTML();
	initBayesClassifier();
	if(readArguments(argc, argv)==-1) exit(-1);
	myData=makeData(judge_file);

	printf("Loading hashes -- be patient!\n");
	loadMassBayesCategories(fbc_dir);
	optimizeFBC(&NBJudgeHashList);

	printf("Classifying\n");
	start=clock();
	mkRegexHead(&myRegexHead, myData, 0);
	removeHTML(&myRegexHead);
	regexMakeSingleBlock(&myRegexHead);
	normalizeCurrency(&myRegexHead);
	regexMakeSingleBlock(&myRegexHead);

//	printf("%ld: %.*ls\n", myRegexHead.head->rm_eo - myRegexHead.head->rm_so, myRegexHead.head->rm_eo - myRegexHead.head->rm_so, myRegexHead.main_memory);

	myHashes.hashes = malloc(sizeof(HTMLFeature) * HTML_MAX_FEATURE_COUNT);
	myHashes.slots = HTML_MAX_FEATURE_COUNT;
	myHashes.used = 0;
	computeOSBHashes(&myRegexHead, HASHSEED1, HASHSEED2, &myHashes);

	classification=doBayesPrepandClassify(&myHashes);
	end=clock();
	printf("Classification took %lf milliseconds\n", (double)((end-start)/(CLOCKS_PER_SEC/1000)));
	printf("Best match: %s prob: %lf pR: %lf\n", classification.primary_name, classification.primary_probability, classification.primary_probScaled);
	if(classification.secondary_name != NULL)
		printf("Best match: %s prob: %lf pR: %lf\n", classification.secondary_name, classification.secondary_probability, classification.secondary_probScaled);

	free(myHashes.hashes);
	freeRegexHead(&myRegexHead);
	free(judge_file);
	free(fbc_dir);
	deinitBayesClassifier();
	deinitHTML();
	return 0;
}
